src/nvapi_gpu.cpp

```diff
@@ -382,11 +382,102 @@ extern "C" {
         if (!nvapiAdapterRegistry->IsAdapter(adapter))
             return ExpectedPhysicalGpuHandle(n);
 
-        *pSize = adapter->GetVirtualVRamSize();
+        *pSize = adapter->GetVirtualVRamSize() / 1024;
 
         return Ok(n);
     }
 
+    NvAPI_Status __cdecl NvAPI_GPU_GetMemoryInfo(NvPhysicalGpuHandle hPhysicalGpu, NV_DISPLAY_DRIVER_MEMORY_INFO* pMemoryInfo) {
+        constexpr auto n = __func__;
+        thread_local bool alreadyLoggedOk = false;
+
+        if (log::tracing())
+            log::trace(n, log::fmt::hnd(hPhysicalGpu), log::fmt::ptr(pMemoryInfo));
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pMemoryInfo == nullptr)
+            return InvalidArgument(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        switch (pMemoryInfo->version) {
+            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_1: {
+                auto pMemoryInfoV1 = reinterpret_cast<NV_DISPLAY_DRIVER_MEMORY_INFO_V1*>(pMemoryInfo);
+                pMemoryInfoV1->dedicatedVideoMemory = adapter->GetVRamSize();
+                pMemoryInfoV1->availableDedicatedVideoMemory = 0; // Not found in DXVK
+                pMemoryInfoV1->systemVideoMemory = adapter->GetDedicatedSystemRamSize();
+                pMemoryInfoV1->sharedSystemMemory = adapter->GetSharedSystemRamSize();
+                break;
+            }
+            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_2: {
+                auto pMemoryInfoV2 = reinterpret_cast<NV_DISPLAY_DRIVER_MEMORY_INFO_V2*>(pMemoryInfo);
+                pMemoryInfoV2->dedicatedVideoMemory = adapter->GetVRamSize();
+                pMemoryInfoV2->availableDedicatedVideoMemory = 0; // Not found in DXVK
+                pMemoryInfoV2->systemVideoMemory = adapter->GetDedicatedSystemRamSize();
+                pMemoryInfoV2->sharedSystemMemory = adapter->GetSharedSystemRamSize();
+                pMemoryInfoV2->curAvailableDedicatedVideoMemory = 0; // Not found in DXVK
+                break;
+            }
+            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_3: {
+                auto pMemoryInfoV3 = reinterpret_cast<NV_DISPLAY_DRIVER_MEMORY_INFO_V3*>(pMemoryInfo);
+                pMemoryInfoV3->dedicatedVideoMemory = adapter->GetVRamSize();
+                pMemoryInfoV3->availableDedicatedVideoMemory = 0; // Not found in DXVK
+                pMemoryInfoV3->systemVideoMemory = adapter->GetDedicatedSystemRamSize();
+                pMemoryInfoV3->sharedSystemMemory = adapter->GetSharedSystemRamSize();
+                pMemoryInfoV3->curAvailableDedicatedVideoMemory = 0; // Not found in DXVK
+                pMemoryInfoV3->dedicatedVideoMemoryEvictionsSize = 0;
+                pMemoryInfoV3->dedicatedVideoMemoryEvictionCount = 0;
+                break;
+            }
```

Comment:
I wonder, can we just use the given pointer as-it without reinterpret-casting it? I think it should be possible because the offset of each member appears to be the same in every struct version in which it's present. That should allow us to make this a bit more DRY, if you don't mind some classic switch fallthrough:

```suggestion
            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_3: 
                pMemoryInfo->dedicatedVideoMemoryEvictionsSize = 0;
                pMemoryInfo->dedicatedVideoMemoryEvictionCount = 0;
                [[fallthrough]];
            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_2:
                pMemoryInfo->curAvailableDedicatedVideoMemory = …;
                [[fallthrough]];
            case NV_DISPLAY_DRIVER_MEMORY_INFO_VER_1:
                pMemoryInfo->dedicatedVideoMemory = adapter->GetVRamSize();
                pMemoryInfo->availableDedicatedVideoMemory = …;
                pMemoryInfo->systemVideoMemory = adapter->GetDedicatedSystemRamSize();                
                pMemoryInfo->sharedSystemMemory = adapter->GetSharedSystemRamSize();
                break;
```

Reply:
yes, this looks cool, we should probably do this for other places too.